HPCC-24058 Add manifest command to esdl tool

asselitx opened the pull request:
The manifest command builds an ESDL service binding or bundle from an input
manifest file. The manifest contains the outline of the binding and can use
Include statements to load external files to include in the output. Supported
Include files are ESDL scripts, XSL transforms for logging and ESDL
definitions.

- Add tests to to exercise the esdl manifest command.
- Refactor the EsdlExample into separate independent examples, including one
  showing the use of the manifest tool.
- Add a README explaining the use of the new command.

The bulk of the changes in this PR are related to updates to the EsdlExample.
So although there are many changed/deleted/moved files, the code changes
for implementing the new command are within the esdl tool project folder.

Signed-off-by: Terrence Asselin <terrence.asselin@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Cloud-compatibility
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [x] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
Added new regression cases in the testing/esp/esdlcmd project.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

Commit:
HPCC-24058 Add manifest command to esdl tool

The manifest command builds an ESDL service binding or bundle from an input
manifest file. The manifest contains the outline of the binding and can use
'Include' statements to load external files to include in the output. Supported
Include files are ESDL scripts, XSL transforms for logging and ESDL
definitions.

- Add tests to to exercise the esdl manifest command.
- Refactor the EsdlExample into separate independent examples, including one
  showing the use of the manifest tool.
- Add a README explaining the use of the new command.

Signed-off-by: Terrence Asselin <terrence.asselin@lexisnexisrisk.com>

afishbeck commented:
I think initfiles/examples/EsdlExample/Manifest needs a copy of RoxieEchoPersonInfo.ecl for the walkthrough to work.

afishbeck commented on initfiles/examples/EsdlExample/CppPlugin/README.md:
````diff
@@ -0,0 +1,82 @@
+
+9. Test calling the roxie service
+```
+soapplus -url http://.:8088/EsdlExample -i cpprequest.xml
````
like above, should be: "../../cpprequest.xml"

asselitx replied:
ok fixed

afishbeck commented on tools/esdlcmd/esdlcmd_common.hpp:
```diff
@@ -142,6 +142,13 @@ typedef IEsdlCommand *(*EsdlCommandFactory)(const char *cmdname);
 #define ESDLOPT_INCLUDE_PATH_INI        "esdlIncludePath"
 #define ESDLOPT_INCLUDE_PATH_USAGE      "   -I, --include-path <include path>    Locations to look for included esdl files\n"
+#define ESDLOPT_RECURSIVE               "--recursive"
+#define ESDLOPT_RECURSIVE_S             "--r"
```
Short names have one dash, long names have two.  So this should be simply "-r".

asselitx replied:
will fix

afishbeck commented on tools/esdlcmd/esdlcmd_core.cpp:
```diff
@@ -1167,13 +1168,166 @@ class Esdl2CppCmd : public EsdlHelperConvertCmd
+        puts("Options:");
+        puts("   --rollup         Output a single XML file containing all definitions used by");
+        puts("                    <serviceName>");
+        puts("   -r, --recursive  Process all includes. Implied when generating monolithic");
```
I see the help is correct for the short version with a single dash.  The define must have been a typo.

asselitx replied:
Yes I think so.

afishbeck commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2020 HPCC Systems®.
```
2022

asselitx replied:
fixed

ghalliday commented:
@asselitx this looks like it has stalled.  Please can you either close, or respond to the comments.

asselitx commented:
@ghalliday I will resume work on this PR later this week.

asselitx commented on initfiles/examples/EsdlExample/BasicRoxie/README.md:
```diff
@@ -0,0 +1,66 @@
+
+## DynamicESDL
+
+You must have configured at least one DynamicESDL ESP Service in this environment.  Either through configmgr or by editing environment.xml. The instance can be bound on port 0, or on the same port as where you're planning to run your service, or both. The purpose of the instance is to provide the configuration needed for your service. If there's an instance configured on the same port, that insta ...
```
I agree that some newer examples including containers will be good, but I'd prefer not to include that in the scope of this ticket. When adding my manifest example to the directory it was difficult to understand how the examples worked, they were intermingled and they didn't have a consistent structure. I worried that just putting a new example in the mix would be unhelpful. 

So I wanted to make them easier to run and understand without changing their basic nature. I will be explicit in the README that these are examples of bare-metal deployments, which I think is useful to have documented. If you'd like I'll also open a ticket to add container and/or esp app-based instructions as either separate examples or updates to these.

asselitx commented on initfiles/examples/EsdlExample/BasicRoxie/README.md:
```diff
@@ -0,0 +1,66 @@
+
+## EclWatch
+
+An instance of EclWatch must be running, and you must have shell access to that node.
```
Good point- trying to be too general is maybe just confusing here. I'll update the readmes to all use the assumption that you are running the stock-installed environment.xml on a single machine that you can access.

asselitx commented on initfiles/examples/EsdlExample/JavaPlugin/EsdlExampleService.java:
```diff
@@ -13,4 +13,3 @@ public JavaEchoPersonInfoResponse JavaEchoPersonInfo(EsdlContext context, JavaEc
         return response;
     }
 }
-
```
Just a mistake - along with the other files missing terminal newlines. I'll fix this and the others.

asselitx commented on initfiles/examples/EsdlExample/Manifest/README.md:
````diff
@@ -0,0 +1,71 @@
+
+4. Create the binding from a manifest file
+```
+esdl manifest manifest_roxie.xml -I EsdlScripts --outfile manifest_example_binding.xml
````
All the examples are named in terms of how the 'backend' is implemented- I guess I was just running on that track. I'll rename it not referencing roxie.

asselitx commented on initfiles/examples/EsdlExample/Manifest/README.md:
````diff
@@ -0,0 +1,71 @@
+esdl bind-service myesp 8088 esdlexample.1 EsdlExample --config manifest_example_binding.xml --overwrite
+```
+
+1. Test calling the roxie service
````
Fixed numbering and adjusted wording, though in keeping with the idea that these examples largely differ by what the backend or implementation is for each.

asselitx commented on tools/esdlcmd/README.md:
````diff
@@ -0,0 +1,424 @@
+</em:Manifest>
+```
+
+The manifest nodes must use the namespace "urn:hpcc:esdl:manifest", and includes elements configuring asingle service.
````
fixed

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,424 @@
+
+The manifest nodes must use the namespace "urn:hpcc:esdl:manifest", and includes elements configuring asingle service.
+
+A `<em:ServiceBinding>` element contains the binding configuration for the service, including:
```
No it is not comprehensive, but new elements could require an update depending on how their contents need to be treated. Any child nodes of `<em:SerivceBinding>` element are copied to output. Any descendant element named *Transform* or *Scripts* are handled specially - their output isn't a copy through, and they can have `<em:Include>` children.

I will need to adjust the code to handle other service-level attributes. I'd plan to handle them like I do now for the `service` and `id` attributes- copy them through to the appropriate output element. This would be good to chat about next week.

Only one `<em:SerivceBinding>` is allowed.

asselitx replied:
After talking with Tony this is the change I made. All attributes from the `ServiceBinding` are copied through to the output `Binding` node. An optional `BindingDefinition` can be the child of `ServiceBinding`, and if present is output as the `Definition` node. Any attributes of `BindingDefinition` are copied through to the output `Definition` node. 

If `BindingDefinition` isn't present, or the required attributes aren't present, the tool will synthesize and output them as it currently does. 

This way you can define attributes in either the `ServiceBinding` or `BindingDefinition` node if you need to, but its not required for you to include a `BindingDefinition` in the manifest if you don't need any custom attributes on the output Definition.

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,424 @@
+Attribute   | Required? | Value  | Usage
+------------|-----------|--------|---------------
+`@id`       | yes       | string | Name of the binding to generate. Will appear in trace log output to identify this binding when loaded.
+`@service`  | yes       | string | Name of the service the generated binding will bind to. This service name must match the ESDL definition's service name.
```
That is a reasonable point, but I'd like to make that change in a future PR.

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,424 @@
+`@id`       | yes       | string | Name of the binding to generate. Will appear in trace log output to identify this binding when loaded.
+`@service`  | yes       | string | Name of the service the generated binding will bind to. This service name must match the ESDL definition's service name.
+
+A `<em:ServiceDefinition>` element contains the ESDL definition of the service.
```
Fixed typo. The tool doesn't prevent you from including multiples but the output would not be sensible. I think it would make sense to enforce only one per manifest file.

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,424 @@
+----------|-----------|--------|---------------
+`@file`   | yes       | string | Name of the file to include. If not an an absolute path, all of the -I include paths are searched to find it
+
+The `<em:Include>` operation can be used in these places in the manifest:
```
Yes the list is comprehensive. I think you have a good point. I'll update so that its supported anywhere, but I'd like to support recursive includes in a future PR.

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,424 @@
+                            Specify the format of the output. Allowed values are
+                            'binding' or 'bundle'. The default is 'bundle' which
+                            includes the ESDL Service Definition (if provided). This
+                            option overrides the Manifest[@outputFormat] in the file.
```
OK yes I see that wording is confusing. I'll update.

asselitx commented on tools/esdlcmd/esdlcmd_shell.cpp:
```diff
@@ -196,6 +196,7 @@ void EsdlCMDShell::usage()
            "   bind-method       Configure method associated with existing ESDL binding.\n"
            "   unbind-method     Remove method associated with existing ESDL binding.\n"
            "   get-binding       Get ESDL binding.\n"
+           "   manifest     Build a service binding or bundle from a manifest file.\n"
```
ok done

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+            }
+
+            #ifdef _MANIFEST_DEBUG_
+            DBGLOG("Binding Output:\n%s", binding.str());
```
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+                while (it != xpp->getNsEnd())
+                {
+                    if (it->first.compare("xml")!=0)
+                        binding.appendf(" xmlns:%s=\"%s\"", it->first.c_str(), it->second);
```
It appears that the xpp namespace iterator does not include the default namespace. That is only be inferred by interrogating the uri of a start tag. So I will need a different method to handle default namespaces.

HPCCSmoketest commented:
This PR (16120) is in the 11/56 position of the queue. Estimated start time is after ~5.10 hour(s)

ghalliday commented:
@asselitx still stalled...

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+            return false;
+        }
+
+        void appendStartTag(IXmlPullParser* xpp, StartTag& stag, bool writeNsAttr = false)
```
Removed default parameter and generalized handling of namespaces.

asselitx commented on initfiles/examples/EsdlExample/BasicRoxie/README.md:
````diff
@@ -0,0 +1,66 @@
+```
+    HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems®.
````
fixed

asselitx commented on initfiles/examples/EsdlExample/BasicRoxie/README.md:
````diff
@@ -0,0 +1,66 @@
+soapplus -url http://.:8088/EsdlExample -i roxierequest.xml
+```
+
+6. Interact with the services by browsing to `http://<DynamicEsdlIP>:8088`
````
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+                throw makeStringExceptionV(ESDL_MANIFEST_ERR, "Unknown resource type (%d) for <Include>, file='%s' on line #%d column #%d", includeType, filename, xpp->getLineNumber(), xpp->getColumnNumber());
+
+            binding.append('\n');
+            // Indent each line of the included file by the current indent level as appending
```
I will address this in another ticket.

asselitx replied:
Added comment to https://track.hpccsystems.com/browse/HPCC-28740

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+            return type;
+        }
+
+        int processInclude(IXmlPullParser* xpp, StartTag& root, IncludeType includeType)
```
Now checking for and escaping cdata markup to prevent nested cdata.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+
+                        lastType = type;
+
+                        #ifdef _MANIFEST_DEBUG_
+                        //UERRLOG("processBinding while:(...%s)", binding.length()<100 ? binding.str() : binding.str()+binding.length()-100);
+                        #endif
```
Fixed these cases.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+
+        void appendStartTag(IXmlPullParser* xpp, StartTag& stag, bool writeNsAttr = false)
+        {
+            // Attributes don't have their ns prefixes copied
```
Fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+                if (!content.length())
+                    throw makeStringExceptionV(ESDL_MANIFEST_ERR, "Unable to load <Include>, file='%s' on line #%d column #%d", filename, xpp->getLineNumber(), xpp->getColumnNumber());
+
+                content.trim();
```
Moved as suggested.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+                    throw makeStringExceptionV(ESDL_MANIFEST_ERR, "Unable to load <Include>, file='%s' on line #%d column #%d", filename, xpp->getLineNumber(), xpp->getColumnNumber());
+
+                content.trim();
+                if (startsWith(content.str(), SCRIPTS_START))
+                {
+                    content.remove(0, strlen(SCRIPTS_START));
+                    if (*(content.str()) && '\n' == *(content.str()))
+                        content.remove(0,strlen("\n"));
+                }
+                if (endsWith(content.str(), SCRIPTS_END))
+                    content.setLength(content.length()-strlen(SCRIPTS_END));
```
Adjusted code to make better assumptions and handle root namespace definitions.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+                if (!content.length())
+                    throw makeStringExceptionV(ESDL_MANIFEST_ERR, "Unable to load <Include>, file='%s' on line #%d column #%d", filename, xpp->getLineNumber(), xpp->getColumnNumber());
+            }
+            else if (INCLUDE_LOG_TRANSFORM == includeType)
```
I've updated the symbol name as suggested. Future update for arbitrary includes may include a refactoring of this method, so I'll address possible removal of the symbol all together at that time.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+                        level--;
+                        break;
+
+                    case XmlPullParser::END_DOCUMENT:
+                        level = 0;
+                        break;
```
Changed all occurrences except at the root level to throw an exception.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+            binding.append("]]>");
+        }
+
+        inline void appendNonWhitespaceContent(IXmlPullParser* xpp)
```
I do refer to placing ESDL Scripts, ESDL Definitions or XSLT inline, but of those only XSLT could have mixed content. On line 123 of the readme I point out that you need to use an `Include` operation to include XSLT that you don't want escaped. 

If we do want more or different support for inline text content or ways of including XML in CDATA I'd like to defer that to a forthcoming PR.

asselitx commented:
@afishbeck and @timothyklemm  could you please review the updates? I didn't rebase to latest or squash yet to help keep the diffs clean. The follow-up ticket for requested changed not addressed now is HPCC-28740.

timothyklemm commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                        break;
+
+                    case XmlPullParser::END_DOCUMENT:
+                        throw makeStringExceptionV(ESDL_MANIFEST_ERR, "Unexpected end of document at line #%d, column #%d", xpp->getLineNumber(), xpp->getColumnNumber());
```
Will we see an unexpected end of document, or will the parser through an exception? The parser definitely throughs exceptions in some situations, and I'm wondering if this is one of them.

asselitx replied:
Tried a few different ways of 'unexpectedly' ending the document and in each case the parser or tokenizer threw an exception before it got to my case. I'll remove these checks.

timothyklemm commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+         * @param root reference to the 'ServiceDefinition' node read
+         * @return int type of the last node read
+         */
+        int processServiceDefinition(IXmlPullParser* xpp, StartTag& root)
```
Can this be renamed to something like `processEsdlServiceDefinition`?

asselitx replied:
sure- or perhaps EsdlDefinition? Either way if I change it should I change the element name to match?

timothyklemm replied:
If a bundle is viewed as a superset of a binding, maybe `processBundleDefinitions` versus `processBindingDefinition`. I wouldn't complain about consistency between the markup and function name.

asselitx replied:
Tony said he'd prefer EsdlDefinition, so I'll change that name for the function and for the manifest namespace element.

timothyklemm commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+            {
+                const char* msg = xe.what();
+                UERRLOG("Error reading manifest file (%s) at line #%d, column #%d: %s", optManifestPath.str(), xe.getLineNumber(), xe.getColumnNumber(), msg ? msg : "Unknown");
+            }
```
Still not clearing `result` on parser exceptions.

asselitx replied:
Nope that was an oversight. Fixed.

timothyklemm commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            catch (XmlPullParserException& xe)
+            {
+                const char* msg = xe.what();
+                UERRLOG("Error reading manifest file (%s) at line #%d, column #%d: %s", optManifestPath.str(), xe.getLineNumber(), xe.getColumnNumber(), msg ? msg : "Unknown");
```
Could be a misleading message since multiple parsers can be used with file inclusion.

asselitx replied:
Removed use of additional parsers.

timothyklemm commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                        {
+                            bindingFileIO->write(0, binding.length(), binding.str());
+                        } else {
+                            UERRLOG("Error writing to binding file (%s).", optOutputPath.str());
```
Ignoring the potential of other artifact types, not all output files are binding files. Better to keep it generic and omit "binding".

timothyklemm commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            }
+            return result;
+        }
+};
```
EOLN

timothyklemm reviewed (changes requested):
This review only covers esdlcmd_manifest.cpp.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+
+        EsdlManifestCmd() {}
+
+        int processCMD() override
```
Is this style? I thought an override of a virtual method was implicitly virtual in all derived classes where its overridden. Core cpp guidelines suggest to use one or the other: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+        EsdlManifestCmd() {}
+
+        int processCMD() override
+        {   bool result = false;
```
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                xpp2->readStartTag(stag);
+                if (streq(stag.getLocalName(), rootName))
+                {
+                    rootStagEndOffset = xpp2->getColumnNumber();
```
this code is removed due to include script simplification.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                if (streq(stag.getLocalName(), rootName))
+                {
+                    rootStagEndOffset = xpp2->getColumnNumber();
+                    if (nullptr == nsList)
```
code removed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            return rootStagEndOffset;
+        }
+
+        int processInclude(IXmlPullParser* xpp, StartTag& root, IncludeType includeType, NamespaceFrame* nsList)
```
I'll look into a refactor with [HPCC-28740](https://track.hpccsystems.com/browse/HPCC-28740). The nsList is removed.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                //    copied through to the output cdata-enclosed 'Scripts' element
+                //    along with all the children (verbatim) of the root node.
+                // 2. An xml fragment containing one or more script elements. All
+                //    all elements are copied verbatim to the cdata-enclosed 'Scripts'
```
that comment is removed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                // 2. An xml fragment containing one or more script elements. All
+                //    all elements are copied verbatim to the cdata-enclosed 'Scripts'
+                //    element of the output.
+                size_t rootStagEndOffset = captureNamespacesInRootNamed(rawFile, "Scripts", nsList);
```
code removed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                    rawFile.remove(0, rootStagEndOffset);
+                    if (*(rawFile.str()) && '\n' == *(rawFile.str()))
+                        rawFile.remove(0,strlen("\n"));
+                    if (endsWith(rawFile.str(), SCRIPTS_END))
```
removed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                    if (*(rawFile.str()) && '\n' == *(rawFile.str()))
+                        rawFile.remove(0,strlen("\n"));
+                    if (endsWith(rawFile.str(), SCRIPTS_END))
+                    rawFile.setLength(rawFile.length()-strlen(SCRIPTS_END));
```
removed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                for (std::string include : includeSearchPaths)
+                    effectiveIncludePath.append(include.c_str()).append(ENVSEPCHAR);
+
+                int effectiveLength = effectiveIncludePath.length();
```
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            }
+            else if (INCLUDE_ESDL == includeType)
+            {
+                StringBuffer effectiveIncludePath;
+
+                for (std::string include : includeSearchPaths)
+                    effectiveIncludePath.append(include.c_str()).append(ENVSEPCHAR);
+
+                int effectiveLength = effectiveIncludePath.length();
+                if (effectiveIncludePath.charAt(effectiveLength) == ENVSEPCHAR)
+                    effectiveIncludePath.setLength(effectiveLength-1);
```
reworked- see above.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                    effectiveIncludePath.append(include.c_str()).append(ENVSEPCHAR);
+
+                int effectiveLength = effectiveIncludePath.length();
+                if (effectiveIncludePath.charAt(effectiveLength) == ENVSEPCHAR)
```
reworked loop and moved to `finalizeOptions()` putting result in member StringBuffer.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            DBGLOG("Output:\n%s", binding.str());
+            #endif
+
+            return 0;
```
changed code to just return error in catch statements.

asselitx replied:
good point- simplified

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            if (iter.done())
+            {
+                usage();
+            } else {
```
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                        result = true;
+                }
+            }
+            return result;
```
good catch, fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            }
+
+            #ifdef _MANIFEST_DEBUG_
+            DBGLOG("Output:\n%s", binding.str());
```
The esdl cmd main application initializes jlog using the stderrLogMsgHandler with no message fields. I'll switch all diagnostic output (errors, warning, progress etc) to use jlog and keep usage() and the final printing of the output to stdout using other non-jlog functions (puts, printf).

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            puts("                        Use once for each path.");
+            puts("    --outfile <filename>");
+            puts("                        Path and name of the output file");
+            puts("    --output-format <format>");
+            puts("                        Specify the format of the output. Allowed values are");
+            puts("                        'binding' or 'bundle'. The default is 'bundle' which");
+            puts("                        includes the ESDL Service Definition (if provided). This");
+            puts("                        option overrides Manifest[@outputFormat] in the file.");
```
Tried to clear up the language.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                    throw makeStringExceptionV(ESDL_MANIFEST_ERR, "Unknown value '%s' provided for --output-format option.", optOutputFormat.get());
+            }
+
+            return true;
```
The other commands are lazily validating paths like these - allowing errors to be thrown/returned at the point of use.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+
+            if (foundFile->isFile() == fileBool::foundYes)
+                return true;
+            else if (isAbsolutePath(filePath))
```
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                {
+                    out.append(in, offset, unsigned(cDataEnd - (in + offset)));
+                    // Ensure we're not looking at a substring of an already encoded cdata end
+                    if (encodedCdataEnd == 0 || (encodedCdataEnd != 0 && encodedCdataEnd > cDataEnd))
```
checked and you were right. Fixed to encode each instance found even if previously encoded.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                    if (encodedCdataEnd == 0 || (encodedCdataEnd != 0 && encodedCdataEnd > cDataEnd))
+                    {
+                        // encode the found ]]>
+                        out.append("]]]]><![CDATA[>");
```
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                if (uri && streq(uri, MANIFEST_URI))
+                    return true;
+                else
+                    UWARNLOG("Found node named '%s' with namespace '%s', but the namespace '%s' may have been intended.", stag.getLocalName(), uri, MANIFEST_URI);
```
fixed- only using non-jlog to output to stdout. Jlog is functional and prints to stderr with no field headers.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+        /**
+         * @brief Process the 'Scripts' node and children.
+         *
+         * The 'Scripts' node contents are enclosed with CDATA and another 'Scripts' node.
+         * Any namespaces defined in the root 'Scripts' node of an included file are
+         * copied to the CDATA-enclosed 'Scripts' node of output.
```
Taken your suggestion to include rooted files as-is.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+            pushNamespaces(xpp, &root);
+            // If an included file defines any namespaces in its root <Scripts> node
+            // then cache those ns definitions in this frame
+            NamespaceFrame nsForEnclosedScripts(newNamespaces.top());
```
removed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                    case XmlPullParser::END_TAG:
+                        xpp->readEndTag(etag);
+
+                        if (isRootEndTagNamed(etag, "Scripts", level))
```
fixed

asselitx replied:
Must have been a neutrino flipping my bits.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                        lastType = type;
+
+                        #ifdef _MANIFEST_DEBUG_
+                        UERRLOG("processBinding while:(...%s)", binding.length()<100 ? binding.str() : binding.str()+binding.length()-100);
```
Changed to DBGLOG

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+        /**
+         * @brief Track namespaces newly defined in the start tag
+         *
+         * Push a frame onto the stack containing any namespaces newly
+         * defined in this start tag. This includes any defaults seen
```
clarified this in the comments and moved all comments about this issue to the top.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+        inline void popNamespaces()
+        {
+            NamespaceFrame newNsForPairedStartTag = newNamespaces.top();
+            for (std::string ns : newNsForPairedStartTag)
```
done

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+            return type;
+        }
+
+        bool processManifest()
```
Added task to HPCC-28740

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,884 @@
+                                    type = processServiceBinding(xpp.get(), stag);
+                                else if (isManifestNodeNamed(stag, MANIFEST_TAG_SVC_DEFINITION))
+                                    type = processServiceDefinition(xpp.get(), stag);
+                                else
```
I'm inclined to say that its treated like any other unrecognized markup and copied through to the output. This afternoon I'll see if Tony has any thoughts if that would be detrimental to future backward compatibility. I'm also not checking in every loop for elements in the manifest namespace in places where they're unsupported or nonsensical.

timothyklemm replied:
I would not expect to see any manifest markup in output that is not meant to be a manifest. Manifest markup is not expected anywhere the output will be used. Between omission of markup or failing, I would probably choose failure because we don't know what the expected output should be.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1409 @@
+                        UWARNLOG("Generating EsdlBundle output without ESDL Service Definition. This may cause runtime issues.");
+
+                    #ifdef _MANIFEST_DEBUG_
+                    UERRLOG("processBinding final:(%s)", binding.str());
```
switched

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+Because the tool is permissive and flexible, copying through all markup not in the manifest namespace, the manifest elements are only required in order to take advantage of the automated processing and simplified format of the manifest file. This example highlights the recommended usage of manifest elements to use the tool's capabilities. Although you could replace some of the elements below with ...
+
+* `<em:Manifest>` Is the required root element. By default the tool outputs a `bundle`, though you may explicitly override that on the command line or by providing an `@outputType=binding` attribute.
+* `em:ServiceBinding>` is valid for both `bundle` and `binding` output. It is necessary to enable recognition of `<em:Scripts>`, and `<em:Transform>` elements.
```
Missing opening `<`.

asselitx replied:
fixed

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+* `<em:Manifest>` Is the required root element. By default the tool outputs a `bundle`, though you may explicitly override that on the command line or by providing an `@outputType=binding` attribute.
+* `em:ServiceBinding>` is valid for both `bundle` and `binding` output. It is necessary to enable recognition of `<em:Scripts>`, and `<em:Transform>` elements.
+* `<em:EsdlDefinition>` is relevant only for `bundle` output. It is necessary to enable element order preservation and recognition of `<em:Include>` as a descendant element.
+* `<em:Include>` causes external file contents to be inserted in place of the element. The processing of included files is context dependent- the parent of the `<em:Include>` element dictates how the file is handled. File inclusion facilitates code reuse in a configuration as code development environment.
```
Add a space between *dependent* and *-*.

asselitx replied:
Well, if we're going to get all Chicago Manual of Style it should probably be either an em dash or semicolon instead of a hyphen anyways. Since we're not doing fancy  typesetting I'll make it a semicolon.

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+* `em:ServiceBinding>` is valid for both `bundle` and `binding` output. It is necessary to enable recognition of `<em:Scripts>`, and `<em:Transform>` elements.
+* `<em:EsdlDefinition>` is relevant only for `bundle` output. It is necessary to enable element order preservation and recognition of `<em:Include>` as a descendant element.
+* `<em:Include>` causes external file contents to be inserted in place of the element. The processing of included files is context dependent- the parent of the `<em:Include>` element dictates how the file is handled. File inclusion facilitates code reuse in a configuration as code development environment.
+* `<em:Scripts>` and `<em:Transform>` trigger preservation of element order for all descendent elements and enable `<em:Include>` recognition. XML element ordering is significant to proper ESDL script, XSLT, and ESXDL content processing. The platform's `IPropertyTree` implementation, used when loading an artifact, does not preserve order. Output configuration files must embed this content as text ...
```
Consider ending the list item after the first sentence and making the remainder a stand-alone paragraph following the list. This is the second bullet point that mentions order preservation but the first that explains its significance.

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+#### Manifest
+Required root element of all manifest files that create output:
+
+- `bundle` output is created by default. It can be made explicit by setting either command line option `--output-type` to `bundle` and/or `@outputType` is `bundle`.
+- `binding` output is created when either command line option `--output-type` is `binding` and/or `@outputType` is `binding`.
```
Suggest a minor update to phrasing - "either command line option ..." can be (mis)interpreted to consider "@outputType" to be a command line option. Maybe "either the command line option ... and/or the manifest property ...".

asselitx replied:
clarified

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+| - | :-: | - | - |
+| `@esdlservice`  | Yes | string | - Name of the ESDL service to which the binding is bound. Output on the `Binding/Definition` element. <br/>- Used to generate a value for `Definition/@id` for `bundle` type output. Note that `Definition/@id` is not output by the tool for `binding` output as it would need to match the `ESDLDefinitionId` passed on command line `esdl bind-service` call, which may d ...
+
+Second are binding or service-specific attributes. This includes any attribute not in the other categories. These are all output on the `Binding/Definition` element. This category will likely grow over time but currently is just `@auth_feature`:
```
Second sentence is "odd". I wouldn't expect to see a reference to "all other" before being presented with at least a name/description of the other categories. It seems like this should be the last category described.

asselitx replied:
Restructured this section a bit with a preamble describing the three categories.

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+| - | :-: | - | - |
+| `@esdlservice`  | Yes | string | - Name of the ESDL service to which the binding is bound. Output on the `Binding/Definition` element. <br/>- Used to generate a value for `Definition/@id` for `bundle` type output. Note that `Definition/@id` is not output by the tool for `binding` output as it would need to match the `ESDLDefinitionId` passed on command line `esdl bind-service` call, which may d ...
+
+Second are binding or service-specific attributes. This includes any attribute not in the other categories. These are all output on the `Binding/Definition` element. This category will likely grow over time but currently is just `@auth_feature`:
```
Last sentence is incorrect. A check of `EsdlServiceImpl::configureTargets` reveals:
- namespace
- returnSchemaLocationOnOK

asselitx replied:
added

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+
+| Attribute   | Required? | Value  | Usage |
+| - | :-: | - | - |
+| `@auth_feature` | No | string | Used declare authorization settings if they aren't present in the ESDL Definition, or override them if they are. Additional documentation on this attribute is forthcoming. |
```
"Used **to** declare ..."

asselitx replied:
fixed

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+The element is not required since it is possible to embed a complete `Definitions` element hierarchy in the manifest.
+
+#### Include
+Optional element that imports the contents of another file into the output, in place of the itself. The outcome of the import depends on the context in which this element is used. See [EsdlDefinition](#esdldefinition), [Scripts](@scripts), and [Transform](#transform) for more information.
```
"... in place of the itself." Remove "the"

asselitx replied:
fixed

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+| - | :-: | - | - |
+| `@file` | Yes | file path | Full or partial path to an external file to be imported. If a partial path is outside of the tool's working directory, the tool's command line must specify the appropriate root directory using either `-I` or `--include-path`. |
+
+>Any XSLTs or ESDL Scripts written inline in a manifest file will have XML escaping applied where required to generate valid XML. If an XLST contains any text content or markup that needs to be preserved as-is (no XML escaping applied) then be sure to use an `<em:Include>` operation. Included files are inserted into the output as-is, with the exception of preventing nested CDATA. If the included  ...
```
XLST -> XSLT

asselitx replied:
fixed

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+Optional repeatable element appearing within an `<em:ServiceBinding>` element that processes child elements and creates output expected for an ESDL binding:
+
+- The `<em:Scripts>` element is replaced on output with `<Scripts>`. Then all content is enclosed in a CDATA section after wrapping it with a new `Scripts` element. That new `<Scripts>` element contains namespaces declared by the input `<em:Scripts>` element. The input `<em:Scripts foo="..." xmlns:bar="..."><!-- content --></em:Scripts>` becomes `<Scripts><![CDATA[<Scripts xmlns:bar="..."><!-- co ...
+- The manifest `<em:Include>` element is recognized to import scripts from external files. The entire file, minus leading and trailing whitespace, is imported.
```
If sentence two is correct, perhaps a warning against including files that include XML declarations.

asselitx replied:
added

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+Optional repeatable element appearing within an `<em:ServiceBinding>` element that processes child elements and creates output expected for an ESDL binding:
+
+- All content is enclosed in a CDATA section. The input `<em:Transform> <!-- content --> </em:Transform>` becomes `<Transform><![CDATA[<!-- content -->]]></Transform>`.
+- The manifest `<em:Include>` element is recognized to import transforms from external files. The entire file, minus leading and trailing whitespace, is imported.
```
ditto

asselitx replied:
added

timothyklemm reviewed (changes requested):
Comments on tools/esdlcmd/README.md only.

timothyklemm reviewed (changes requested):
Just esdlcmd_manifest.cpp ...

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,485 @@
+
+### Manifest File
+
+A manifest file is an XML-formatted template combining elements in and outside of the manifest's `urn:hpcc:esdl:manifest` namespace. Elements of this namespace control the tool while all other markup is copied to the output. The goal of using a manifest file with the tool is to make configuring and deploying services easier:
```
clarified

asselitx commented on tools/esdlcmd/README.md:
````diff
@@ -0,0 +1,485 @@
+    </em:EsdlDefinition>
+</em:Manifest>
+```
+Because the tool is permissive and flexible, copying through all markup not in the manifest namespace, the manifest elements are only required in order to take advantage of the automated processing and simplified format of the manifest file. This example highlights the recommended usage of manifest elements to use the tool's capabilities. Although you could replace some of the elements below with ...
````
clarified

asselitx commented on tools/esdlcmd/README.md:
````diff
@@ -0,0 +1,485 @@
+```
+Because the tool is permissive and flexible, copying through all markup not in the manifest namespace, the manifest elements are only required in order to take advantage of the automated processing and simplified format of the manifest file. This example highlights the recommended usage of manifest elements to use the tool's capabilities. Although you could replace some of the elements below with ...
+
+* `<em:Manifest>` Is the required root element. By default the tool outputs a `bundle`, though you may explicitly override that on the command line or by providing an `@outputType=binding` attribute.
````
lowercased

asselitx commented on tools/esdlcmd/README.md:
````diff
@@ -0,0 +1,485 @@
+```
+Because the tool is permissive and flexible, copying through all markup not in the manifest namespace, the manifest elements are only required in order to take advantage of the automated processing and simplified format of the manifest file. This example highlights the recommended usage of manifest elements to use the tool's capabilities. Although you could replace some of the elements below with ...
+
+* `<em:Manifest>` Is the required root element. By default the tool outputs a `bundle`, though you may explicitly override that on the command line or by providing an `@outputType=binding` attribute.
````
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1311 @@
+
+                //First parameter's order is fixed.
+                const char *arg = iter.query();
+                if (*arg != '-')
```
I'm really just copying the behavior of other esdl commands here. If we're going to update this I think a command-wide change under another ticket would be best.

afishbeck replied:
Congratulations, not supporting manifest files that start with '-' has just replaced the previous "least of my worries". ;)

timothyklemm replied:
They can't all be show stoppers, and I've never asserted this to be one.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1311 @@
+            if (isAbsolutePath(filePath))
+                return false;
+
+            for (std::string include : includes)
```
fixed

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1311 @@
+            }
+            while(type != XmlPullParser::END_TAG && type != XmlPullParser::END_DOCUMENT);
+            EndTag etag;
+            xpp->readEndTag(etag);
```
Probably some undefined behavior. Fixed.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1311 @@
+
+            if (outputType != ManifestType::bundle)
+            {
+                UWARNLOG("The requested binding output will not include the ESDL Definition found in the Manifest.");
```
For the second case what is the operative difference from the first? Is it the difference between using the `@outputType=binding` attribute instead of passing it on the CLI?

timothyklemm replied:
In the first case, the manifest is declared to be capable of producing a bundle. It can explicitly also be used to produce a binding, which is a subset of a bundle. I don't see a need to warn the user that the user has explicitly requested a proper subset of the template.

In the second case, the manifest is explicitly declared to produce a binding. There should be no expectation that an ESDL definition is present and a valid bundle can be output. If a definition is present, will its inclusion in binding output be valid? Should a definition be silently ignored in binding output? Should a user be allowed to request a bundle from a binding template? These are conditions that may warrant generating warning or error messages.

I was not, and am not, suggesting that alternate messaging is necessary now. I was just making the point that there are other, more worthy, candidates for warning and error messages than this.

asselitx commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1311 @@
+                rawFile.loadFile(file);
+                // Trim to ensure correct indentation
+                rawFile.trim();
+                if (!rawFile.length())
```
I can see your point but wouldn't it be more harmful to silently output a binding missing scripts because you forgot to fill in the empty files? I'd think in general you'd expect included files to have some content more often than not.

timothyklemm replied:
I would expect empty files to be less common. I'm just not convinced that an empty file should result in automatic failure. A warning message might be sufficient.

asselitx replied:
The ESDL helper class already returns throws if the file is empty, however, this exception I'm throwing would be after the ESDLHelper has processed the file which should result in a valid defn. So if it is empty at that point I think it is an indication of some problem worthy of an exception.

afishbeck commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1310 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2022 HPCC Systems®.
```
It's now 2023

asselitx replied:
Yes

afishbeck commented on tools/esdlcmd/esdlcmd_manifest.cpp:
```diff
@@ -0,0 +1,1310 @@
+                    if (ManifestType::bundle == outputType)
+                        appendEndTag(MANIFEST_TAG_BUNDLE);
+                    if (ManifestType::bundle == outputType && !foundEsdlDefn)
+                        UWARNLOG("Generating EsdlBundle output without ESDL Definition. This may cause runtime issues.");
```
Perhaps this should be an error?  I think all bundles should have interfaces defined?  If in the future we wanted to make it a warning we could.  Sometimes starting out strict is good.

asselitx replied:
I'm fine with that. I'll plan to throw an exception to prevent the creation of a bundle without the interface- unless you meant just to use UERRLOG?

afishbeck replied:
Throw an exception, we can always relax it later if it is really needed.

asselitx replied:
Fixed in today's commit.

asselitx commented:
@timothyklemm and @afishbeck I believe I've addressed all outstanding comments. Please check latest commit. I'll rebase and squash onto 9.2.x when you think it looks good.

asselitx commented:
@timothyklemm and @afishbeck I've rebased to candidate-9.2.x and fixed the README error in grammar I'd missed. I feel the changes are safe in 9.2 because they're limited to 

1. The initfiles/examples
2. esdl command tests
3. new `manifest` esdl command
4. adding support for `esdl xml` to run directly from the esdl app instead of a separate executable

timothyklemm commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,506 @@
+
+#### Bundle
+
+The bundle includes all the components of the service binding in addition to the ESDL definition.
```
Consider rephrasing. Between this and 426, seems somewhat circular.

asselitx replied:
Removed that wording and instead called back to the usage for binding/bundle around line 45.

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -108,22 +114,38 @@ Required root element of all manifest files that create output:
 | - | :-: | - | - |
 | `@auth_feature` | No | string | Used declare authorization settings if they aren't present in the ESDL Definition, or override them if they are. Additional documentation on this attribute is forthcoming. |
+| `@returnSchemaLocationOnOK` | No | Boolean | When true, a successful SOAP response (non SOAP-Fault) will include the schema location property. False by default. |
+| `@namespace` | No | string | String specifying the namespace for all methods in the binding. May contain variables that are replaced with their values: <br/>- `${service}` : lowercase service name <br/>- `${esdl-service}` : service name, possibly mixed case <br/>- `${method}` : lowercase method name <br/>- `${esdl-method}` : method name, possibly mixed case <br/>- `${optionals}` : comma-delimit ...
```
makes sense, fixed

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,506 @@
+| - | :-: | - | - |
+| `@auth_feature` | No | string | Used to declare authorization settings if they aren't present in the ESDL Definition, or override them if they are. Additional documentation on this attribute is forthcoming. |
+| `@returnSchemaLocationOnOK` | No | Boolean | When true, a successful SOAP response (non SOAP-Fault) will include the schema location property. False by default. |
+| `@namespace` | No | string | String specifying the namespace for all methods in the binding. May contain variables that are replaced with their values: <br/>- `${service}` : lowercase service name <br/>- `${esdl-service}` : service name, possibly mixed case <br/>- `${method}` : lowercase method name <br/>- `${esdl-method}` : method name, possibly mixed case <br/>- `${optionals}` : comma-delimit ...
```
Clarified the ESP replaces them during runtime.

asselitx commented on tools/esdlcmd/README.md:
```diff
@@ -0,0 +1,506 @@
+| `@port`        | No | string | Port on the ESP Process listening for connections to the binding. |
+| `@publishedBy` | No | string | Userid of the person publishing the binding |
+
+The `<em:ServiceBinding>` element is not required since it is possible to embed a complete `Binding` element hierarchy in the manifest.
```
Yeah seems like that got shuffled during edits. I've reworded slightly and moved up.

timothyklemm commented:
I would approve if I could find an option to do so in the phone app.

asselitx commented:
@ghalliday approved for merge and squashed

asselitx commented:
Rebased to target branch again in an attempt to re-trigger the checks that had stalled out.

asselitx commented:
@ghalliday this should be ready now for your final approval and merge. Let me know if I need to do anything else.
